nav2_costmap_2d/plugins/costmap_filters/speed_filter.cpp

```diff
@@ -0,0 +1,343 @@
+      "Filter mask was not received");
+    return;
+  }
+
+  double mask_robot_x, mask_robot_y;  // robot coordinates in mask frame
+
+  if (mask_frame_ != global_frame_) {
+    // Filter mask and current layer are in different frames:
+    // Prepare transform from current layer frame (global_frame_) to mask frame
+    geometry_msgs::msg::TransformStamped transform;
+    tf2::Transform tf2_transform;
+    tf2_transform.setIdentity();  // initialize by identical transform
+    try {
+      transform = tf_->lookupTransform(
+        mask_frame_, global_frame_, tf2::TimePointZero,
+        transform_tolerance_);
+    } catch (tf2::TransformException & ex) {
+      RCLCPP_ERROR(
+        node->get_logger(),
+        "SpeedFilter: failed to get costmap frame (%s) "
+        "transformation to mask frame (%s) with error: %s",
+        global_frame_.c_str(), mask_frame_.c_str(), ex.what());
+      return;
+    }
+    tf2::fromMsg(transform.transform, tf2_transform);
+
+    // Transform (pose.x, pose.y) point from global_frame_
+    // to (mask_robot_x, mask_robot_y) point in mask_frame_
+    tf2::Vector3 point(pose.x, pose.y, 0);
+    point = tf2_transform * point;
+    mask_robot_x = point.x();
+    mask_robot_y = point.y();
+  } else {
+    // Filter mask and current layer are in the same frame:
+    // Just use pose coordinates
+    mask_robot_x = pose.x;
+    mask_robot_y = pose.y;
+  }
+
+  // Converting (mask_robot_x, mask_robot_y) robot position
+  // to filter_mask_ indexes (mask_robot_i, mask_robot_j)
+  double origin_x = filter_mask_->info.origin.position.x;
+  double origin_y = filter_mask_->info.origin.position.y;
+  double resolution = filter_mask_->info.resolution;
+  unsigned int size_x = filter_mask_->info.width;
+  unsigned int size_y = filter_mask_->info.height;
+
+  if (mask_robot_x < origin_x || mask_robot_y < origin_y) {
+    // Coordinates are out of mask bounds. Returning.
+    return;
+  }
+
+  unsigned int mask_robot_i = std::round((mask_robot_x - origin_x) / resolution);
+  unsigned int mask_robot_j = std::round((mask_robot_y - origin_y) / resolution);
+  if (mask_robot_i >= size_x || mask_robot_j >= size_y) {
+    // Coordinates are out of mask bounds. Returning.
+    return;
+  }
+
+  // Getting filter_mask data from cell where the robot placed and
+  // calculating speed limit value
+  int8_t data = filter_mask_->data[mask_robot_j * size_x + mask_robot_i];
+  if (data != nav2_util::OCC_GRID_FREE && data != nav2_util::OCC_GRID_UNKNOWN) {
+    if (percentage_) {
+      // Speed limit is expressed in % from maximum speed of robot
+      speed_limit_ = 1.0 - static_cast<double>(data) / nav2_util::OCC_GRID_OCCUPIED;
+    } else {
+      // Speed limit is expressed in absolute value
+      speed_limit_ = data * multiplier_ + base_;
+    }
+  } else {
+    speed_limit_ = NO_SPEED_LIMIT;
+  }
+
+  if (speed_limit_ != speed_limit_prev_) {
+    if (speed_limit_ != NO_SPEED_LIMIT) {
+      if (percentage_) {
+        RCLCPP_INFO(
```

Comment:
We might to make these debug, since it could be running at costmap update rate (5-10hz) and in that case it would just spam the console.

Reply:
This message appears only once when speed limit was updated, not each time when `process()` is called. So, I do not see the reason make it be RCLCPP_DEBUG.

Reply:
If there's a gradient of speeds (which we foresaw and why we implemented this as a mask rather than a semantics vectorized area), then it changes every pixel it goes over. This could easily spam things except when there are discrete boundary changes. I agree for discrete cases, an INFO would be helpful, but there needs to be something to balance it out so we're not spamming the console.

Reply:
I did not thought about gradient speeds. This might represent an acceleration/deceleration we want if robot to change speed smoothly. I will change message to DEBUG with that case in mind.

Reply:
Done: https://github.com/ros-planning/navigation2/pull/2074/files#diff-d344e376ecbc8e5ecd5045731e13855c6d2a9a955cdaee9b31dddbbee7681466R320